blob/service.go

```diff
@@ -347,127 +353,193 @@ func (s *Service) retrieve(
 		return nil, nil, err
 	}
 
+	eds, err := s.shareGetter.GetEDS(ctx, header)
+	if err != nil {
+		return nil, nil, err
+	}
 	headerGetterSpan.SetStatus(codes.Ok, "")
 	headerGetterSpan.AddEvent("received eds", trace.WithAttributes(
 		attribute.Int64("eds-size", int64(len(header.DAH.RowRoots)))))
 
-	rowIndex := -1
+	// find the index of the row where the blob could start
+	inclusiveNamespaceStartRowIndex := -1
 	for i, row := range header.DAH.RowRoots {
 		if !namespace.IsOutsideRange(row, row) {
-			rowIndex = i
+			inclusiveNamespaceStartRowIndex = i
 			break
 		}
 	}
+	if inclusiveNamespaceStartRowIndex == -1 {
+		return nil, nil, ErrBlobNotFound
+	}
 
-	getCtx, getSharesSpan := tracer.Start(ctx, "get-shares-by-namespace")
+	// end exclusive index of the row root containing the namespace
+	exclusiveNamespaceEndRowIndex := inclusiveNamespaceStartRowIndex
+	for _, row := range header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:] {
+		if namespace.IsOutsideRange(row, row) {
+			break
+		}
+		exclusiveNamespaceEndRowIndex++
+	}
+	if exclusiveNamespaceEndRowIndex == inclusiveNamespaceStartRowIndex {
+		return nil, nil, fmt.Errorf("couldn't find the row index of the namespace end")
+	}
```

Comment:
Q: so, it's ok when we just iterate over `header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:]` till the end. This means that we cover all ranges, am I right?

Reply:
yes, the namespaces are ordered